src/platform/i386/kernel.c

```diff
@@ -164,21 +166,47 @@ kmain(struct multiboot *mboot, u32_t mboot_magic, u32_t esp)
 	/* uses virtual address for VGA. should be after paging_init() */
 	vga_high_init();
 #endif
-	kern_boot_comp();
-	smp_init();
+	kern_boot_comp(0);
 	timer_init();
+	lapic_init();
 	lapic_timer_init();
+	smp_init(cores_ready);
+	bsp_ready = 1;
 	kern_boot_upcall();
+
 	/* should not get here... */
 	khalt();
 }
 
+void
+smp_kmain(void)
+{
+	volatile int cpu_id = get_cpuid();
+	struct cos_cpu_local_info *cos_info = cos_cpu_local_info();
+	printk("Initialize CPU %d\n", cpu_id);
+	tss_init(cpu_id);
+	gdt_init(cpu_id);
+	idt_init(cpu_id);
+
+	paging_init();
+	kern_boot_comp(cpu_id);
+	lapic_init();
+	lapic_timer_init();
+
+	printk("New CPU %d Booted\n", cpu_id);
+	cores_ready[cpu_id] = 1;
+	// waiting for all cored booted
```

Comment:
comment style for single line as per CSG: `/* comment */`.

Reply:
dito